cpp/pagerank_module/pagerank/pagerank.cpp

```diff
@@ -0,0 +1,161 @@
+
+namespace pagerank {
+
+PageRankGraph::PageRankGraph(
+    const uint32_t number_of_nodes, const uint32_t number_of_edges,
+    const std::vector<std::pair<uint32_t, uint32_t>> &edges)
+    : node_count_(number_of_nodes), edge_count_(number_of_edges),
+      out_degree_(std::vector<uint32_t>(number_of_nodes)) {
+  utils::AdjacencyList<uint32_t> in_neighbours;
+  in_neighbours.Init(number_of_nodes);
+  for (const auto [from, to] : edges) {
+    // Because PageRank needs a set of nodes that point to a given node.
+    in_neighbours.AddAdjacentPair(from, to);
+    out_degree_[from] += 1;
+  }
+  for (size_t node_id = 0; node_id < number_of_nodes; node_id++) {
+    const auto &adjacent_nodes = in_neighbours.GetAdjacentNodes(node_id);
+    for (const auto &adjacent_node : adjacent_nodes) {
+      ordered_edges_.emplace_back(adjacent_node, node_id);
+    }
+  }
+}
+
+uint32_t PageRankGraph::GetNodeCount() const { return node_count_; }
+
+uint32_t PageRankGraph::GetEdgeCount() const { return edge_count_; }
+
+const std::vector<std::pair<uint32_t, uint32_t>> &
+PageRankGraph::GetOrderedEdges() const {
+  return ordered_edges_;
+}
+
+uint32_t PageRankGraph::GetOutDegree(const uint32_t node_id) const {
+  return out_degree_[node_id];
+}
+
+void CalculateOptimalBorders(const PageRankGraph &graph,
+                             const int number_of_threads,
+                             std::vector<uint32_t> *borders) {
+  for (int i = 0; i <= number_of_threads; i++)
+    borders->push_back(static_cast<uint64_t>(i) * graph.GetEdgeCount() /
+                       number_of_threads);
+}
+
+void ThreadPageRankIteration(
+    const PageRankGraph &graph, const std::vector<double> &old_rank,
+    const uint32_t lo, const uint32_t hi,
+    std::promise<std::vector<double>> new_rank_promise) {
+  std::vector<double> new_rank;
+
+  for (size_t edge_id = lo; edge_id < hi; edge_id++) {
+    const auto [source, target] = graph.GetOrderedEdges()[edge_id];
+    while (new_rank.size() < target - graph.GetOrderedEdges()[lo].second + 1)
+      new_rank.push_back(0);
+    new_rank.back() += old_rank[source] / graph.GetOutDegree(source);
+  }
+  new_rank_promise.set_value(new_rank);
+}
+
+void AddCurrentBlockToRankNext(const PageRankGraph &graph,
+                               const double damping_factor,
+                               const std::vector<double> &block,
+                               const size_t cluster_id,
+                               const std::vector<uint32_t> &borders,
+                               std::vector<double> *rank_next) {
+  const uint32_t lo = borders[cluster_id];
+  if (lo < graph.GetEdgeCount())
+    while (rank_next->size() <= graph.GetOrderedEdges()[lo].second)
+      rank_next->push_back((1.0 - damping_factor) / graph.GetNodeCount());
+  for (size_t i = 0; i < block.size(); i++) {
+    if (i > 0)
+      rank_next->push_back((1.0 - damping_factor) / graph.GetNodeCount());
+    rank_next->back() += damping_factor * block[i];
```

Comment:
Not sure if I understand everything algorithm wise but shouldn't this be rank_next[i] += damping_factor * block[i] and push back on rank_next until it's at least the same size as the rank_next? :thinking:

Reply:
Yep, this was completely broken. Implemented the simplest possible version. I think performance could be improved, but will further test that.